### DISCUSSION TITLE
Working with Turbo Stream helper?

### DISCUSSION THREAD

**Discussion by Bumppoman** [yippee-fun/phlex#303] [Help] [answered]
Hi,

I'm wondering what you might recommend for working with the turbo-stream helper in rails.  It's more complicated than turbo-frames because the helper is doing a bunch of additional work, not just wrapping the content in a simple tag.

I appreciate any insights!

**Comment by dhnaranjo** [comment]
I'mma just pop in a lil simplified snippet of what I'm doing to use a turbo stream helper. Happy to discuss if you got Qs.

[Here's a related issue](https://github.com/joeldrapper/phlex/issues/273) that talks a bit bout the need for the `view_context` stuff.

```
class UpdateThingJob < ApplicationJob
  def perform(thing)
    thing.make_ready!

    view_context = ThingsController.new.view_context
    thing.broadcast_replace_to(
      Current.account,
      :thing_status,
      html: Views::Things::Show::Ready.new(thing).call,
      view_context:,
      target: view_context.dom_id(thing)
    )
  end
end

module Views
  module Things
    class Show < ApplicationView
      def initialize(thing)
        @thing = thing
      end

      def template
        raw helpers.turbo_stream_from(Current.account, :thing_status)
        turbo_frame id: helpers.dom_id(@thing), target: "_top" do
          if @thing.ready?
            ready
          else
            not_ready
          end
        end
      end

      def not_ready
        # Not ready state content
      end

      class Ready < Show
        def template
          # Ready state content
        end
      end
    end
  end
end
```

**Comment by joeldrapper** [comment]
Hey @Bumppoman. I haven't spent much time on this problem yet, but it's definitely a priority for the library. Have you any thoughts on how it might look?

**Reply by Bumppoman** [reply]
If you have enough time to look and point me in the right direction, I would absolutely put something together.  I spent a day on it earlier this week but nothing seemed promising.  I tried `raw` everywhere primarily.  I could get content to print and even be sent to the browser but not in the order Turbo expects.

Also while I have you here, I'm loving the Phlex experience.  My slowest view is 3x faster than ERB templates and some are as much as 10x faster.  I'm excited to see where things are going and help if possible.

**Reply by joeldrapper** [reply]
So happy to hear that! Want to book some time to pair on this for a bit? https://savvycal.com/joeldrapper/pair

**Reply by Bumppoman** [reply]
I will try to find some time that matches up!

**Comment by joeldrapper** [comment, marked as answer]
Hey @Bumppoman I added support for rendering Phlex components as turbo stream content in turbo-rails 1.4. https://github.com/hotwired/turbo-rails/pull/433

You can use it like this

```ruby
turbo_stream.append "notifications", NotificationComponent.new(@notification)
```

**Reply by mhenrixon** [reply]
Hello again, @joeldrapper 🙈

Is it possible to skip the turbo stream view, e.g., `create.turbo_stream.erb`? I failed to do this in the controller. I figured it was best to ask rather than go crazy.

**Reply by davekaro** [reply]
I was able to make this work three different ways:

First, by manually rendering the component:

```
render turbo_stream: Component.new.call(view_context: view_context)
```

Then, by letting `render` call `render_in` on our passed in object. Here the default content type is text/html so I had to adjust. Also, by default, when a rendering mode is specified (e.g. `turbo_stream:` above), no layout template is rendered. So in this example, I also had to skip the layout.

```
self.content_type = Mime[:turbo_stream]
render Component.new, layout: false
```

Finally, using a `respond_to` block which sort of combines both:

```
respond_to do |format|
  format.turbo_stream { render Component.new, layout: false }
end
```

The last two feel best to me in most cases. I tried adding a custom renderer (https://api.rubyonrails.org/classes/ActionController/Renderers.html#method-c-add) like turbo-rails does, but couldn't get it to work.
